RecoLocalTracker/SiPhase2Clusterizer/test/ClustersValidation.cc

```diff
@@ -163,167 +165,177 @@ void Phase2TrackerClusterizerValidation::analyze(const edm::Event& event, const
-            if (mType == TrackerGeometry::ModuleType::Ph2PSP) {
-                histogramLayer->second.localPosXY[1]->Fill(localPosClu.x(), localPosClu.y());
-                histogramLayer->second.globalPosXY[1]->Fill(globalPosClu.x(), globalPosClu.y());
-                histogramLayer->second.clusterSizePixel->Fill(clustIt->size());
-                ++nClustersPixel;
-            }
-            // Strip module
-            else if (mType == TrackerGeometry::ModuleType::Ph2PSS || mType == TrackerGeometry::ModuleType::Ph2SS) {
-                histogramLayer->second.localPosXY[2]->Fill(localPosClu.x(), localPosClu.y());
-                histogramLayer->second.globalPosXY[2]->Fill(globalPosClu.x(), globalPosClu.y());
-                histogramLayer->second.clusterSizeStrip->Fill(clustIt->size());
-                ++nClustersStrip;
-            }
-
-            /*
-             * Digis related variables
-             */
-
-            std::vector< unsigned int > clusterSimTrackIds;
-
             // Get all the simTracks that form the cluster
+            std::vector<unsigned int> clusterSimTrackIds;
             for (unsigned int i(0); i < clustIt->size(); ++i) {
                 unsigned int channel(Phase2TrackerDigi::pixelToChannel(clustIt->firstRow() + i, clustIt->column())); 
-                unsigned int simTrackId(getSimTrackId(pixelSimLinks, detId, channel));
-                clusterSimTrackIds.push_back(simTrackId);
+                std::vector<unsigned int> simTrackIds(getSimTrackId(pixelSimLinks, detId, channel));
+                for (unsigned int i=0; i<simTrackIds.size(); ++i) {
+                  bool add = true;
+                  for (unsigned int j=0; j<clusterSimTrackIds.size(); ++j) {
+		    // only save simtrackids that are not present yet
+                    if (simTrackIds.at(i)==clusterSimTrackIds.at(j)) add = false;
+                  }
+                  if (add) clusterSimTrackIds.push_back(simTrackIds.at(i));
+		}
             }
 
+	    // find the closest simhit
+	    // this is needed because otherwise you get cases with simhits and clusters being swapped
+	    // when there are more than 1 cluster with common simtrackids
+	    const PSimHit * simhit = 0; // bad naming to avoid changing code below. This is the closest simhit in x
+	    float minx=10000;
+            for (unsigned int simhitidx = 0; simhitidx < 2; ++simhitidx) { // loop over both barrel and endcap hits
+              for (edm::PSimHitContainer::const_iterator simhitIt(simHitsRaw[simhitidx]->begin()); simhitIt != simHitsRaw[simhitidx]->end(); ++simhitIt) {
+                if (rawid == simhitIt->detUnitId()) {
+                  //std::cout << "=== " << rawid << " " << &*simhitIt << " " << simhitIt->trackId() << " " << simhitIt->localPosition().x() << " " << simhitIt->localPosition().y() << std::endl;
+		  if (std::find(clusterSimTrackIds.begin(), clusterSimTrackIds.end(), simhitIt->trackId()) != clusterSimTrackIds.end()) {
```

Comment:
using `std::find()` in a double-nested loop seems like a bad idea. Could the `clusterSimTrackIds` be kept sorted?